### ISSUE TITLE
Keep bundling binaryen.js/wasm.js?

### ISSUE THREAD

**Issue by kripken** [WebAssembly/binaryen#1571]
We currently bundle them in this repo, under `bin/`. I worry though that over time it increases the size of the git repo significantly. Am I overreacting?

Another option would be a separate location for the builds, perhaps.

Another benefit of no longer bundling them is that PRs no longer need to bother with rebuilding&bundling. But on the other side, not bundling them means if a user just wants those `.js` files then they must go someplace else to get them.

**Comment by kripken** [comment]
The builds depend on the compiler too, the emsdk, yeah.

Currently we build a debug version on travis and run tests on that, to prevent regressions. A release build takes quite some time but should also be doable in the 50 minutes travis allows us, so perhaps we could set it to publish to github releases. I'm not really sure how github releasees work, but if new tags (for new versions) could trigger such builds being added, that sounds good.

**Comment by xtuc** [comment]
Yes a tag can trigger a build and then a simple curl (plus a token) can create the release with the files.

**Comment by kripken** [comment]
Thanks @xtuc , this sounds doable then.

**Comment by kripken** [comment]
@dcodeIO how do you handle builds for binaryen.js? In this issue we're considering stopping the bundling in the binaryen repo.

**Comment by dcodeIO** [comment]
My [buildbot](https://github.com/AssemblyScript/binaryen.js/blob/master/.travis.yml) basically

1. Installs and compiles Emscripten incoming from source (using ccache so it's relatively fast)
2. Updates Binaryen, which is a submodule, to latest (for nightlies) or a specific tag if there is a new one (for releases)
3. Runs a [build script](https://github.com/AssemblyScript/binaryen.js/blob/master/scripts/build.js) very similar to `build-js.sh` and outputs the result to `index.js`
4. Runs a [sanity test](https://github.com/AssemblyScript/binaryen.js/blob/master/tests/index.js) to make sure that the binary is working
5. Creates a Git tag, pushes the changes back to GH and publishes to npm

While this doesn't create any GH releases (and index.js is a asm.js binary), this essentially looks, feels and behaves like a simple, regular node module repo.

**Comment by xtuc** [comment]
That's a good point, I would prefer publishing on npm instead of a GH release. It's more convenient.

**Comment by kripken** [comment]
Oh cool! So maybe we can do something like this:

 * Point people that want builds to npm. Maybe most people are looking there anyhow :)
 * Keep the current debug build tests we have (that is, do a quick debug build of binaryen.js and verify it).
 * The one tricky bit is the rest of the tests, that is, that tested the bundled optimized build. The value there is that (rarely) debug builds miss something. The problem is that if a PR here updates the code, there will not be an optimized build for that on npm. Maybe it's just find to drop this test mode, I am thinking.
 * About wasm.js, we currently don't test it here actually, but emscripten does. But over there, we can a tagged build from npm, I think. (Alternatively, we could do a quick debug build and test with that.)

**Comment by dcodeIO** [comment]
> The one tricky bit is the rest of the tests, that is, that tested the bundled optimized build.

Maybe the JS specific tests could then be run by the buildbot (the sanity tests it currently has are relatively basic so any improvement there would be great anyway). One thing to note afaik is that the JS tests cover some parts of the C-API that the C tests do not cover yet.

**Comment by kripken** [comment]
@dcodeIO makes sense, yeah, improving the buildbots test could offset removing tests here.

Another thing I realized meanwhile is that wasm.js can go away once we finish wasm2js - that is meant to be a drop-in replacement for wasm.js. So we'll need to figure out testing for that anyhow, and can remove wasm.js in the meanwhile.

**Comment by kripken** [comment]
Is there a way to link directly to binaryen.js on npm? Maybe there's an obvious place but I can't seem to find it. I think that would be useful for people that just want the binaryen.js build, without npm (might not be many people, but still).

**Comment by xtuc** [comment]
You can use Unpkg

**Comment by dcodeIO** [comment]
It's basically the `index.js` file, i.e.

master: https://raw.githubusercontent.com/AssemblyScript/binaryen.js/master/index.js
version_48: https://raw.githubusercontent.com/AssemblyScript/binaryen.js/v48.0.0/index.js

Also works with rawgit.com as a CDN.

**Comment by kripken** [comment]
And I guess the same question for @xtuc about bundle.run? Just curious how mainstream all these services are...

**Comment by xtuc** [comment]
Yes, millions of peoples rely on these services even if none of them are official products. 

Could we fix the latest builds of binaryen.js? For example https://travis-ci.org/AssemblyScript/binaryen.js/builds/395478938 (I have trouble understanding the log on mobile sorry)

**Comment by dcodeIO** [comment]
From what I can tell unpkg.com is relatively popular among node.js developers because it pretty much behaves like npm with its urls. It's open source and according to its [README](https://github.com/unpkg/unpkg.com) sponsored by CloudFlare and Heroku.

Not sure how common RawGit is, but I like its sheer simplicitly (just [something simple serving GH repos via CloudFlare](https://github.com/rgrove/rawgit) afaik) and that one can pick exact commits as well.

**Comment by kripken** [comment]
I see, thanks @xtuc & @dcodeIO 

That binaryen.js CI error looks like
```
error: failure to process js library "library.js": Missing C define TIME_UTC! If you just added it to struct_info.json, you need to ./emcc --clear-cache
```
Should be fixed by clearing the cache. I assume the bot uses latest incoming? If it does, might be worth clearing the cache on each run (but will add some time to building libc etc.). Or, getting the latest tagged version would avoid that issue too.

**Comment by dcodeIO** [comment]
Commit above should fix that, I hope.

**Comment by dcodeIO** [comment]
> Or, getting the latest tagged version would avoid that issue too.

Being able to just `emsdk install latest` would be ideal, but last time I checked precompiled binaries didn't work on Travis CI because of a glibc mismatch. I think there are a couple issues I opened or replied to at emsdk / emscripten somewhere. Do you know if there has been any progress on making proper static binaries?

**Comment by kripken** [comment]
The binaries work on 16.04, this is what we use for emscripten CI on travis: https://github.com/kripken/emscripten/blob/incoming/Dockerfile Also works on circle (also using 16.04).

**Comment by dcodeIO** [comment]
I see, thanks. *goes learn docker*

**Comment by tlively** [comment]
We stopped bundling built binaryen.js artifacts, so this is resolved.
